Replace libgit2 with gitoxide

blinxen opened the pull request:
As discussed in https://github.com/sharkdp/bat/discussions/3151, I have prepared a PR to replace the `libgit2` usage in `bat` with `gitoxide`. The PR is not finished *yet* because some tests are failing since the produced `git diff` does not match the [exact behavior of `libgit2`](https://github.com/pascalkuthe/imara-diff/issues/21).

I see here two solutions:

* Extend the `imara-diff` crate to generate `git` like diffs
* Adapt the tests to accept the *new* diff format (it is not a format but more like a different diffing algorithm)

What do the maintainers think?

CC: @byron

blinxen committed:
Replace libgit2 with gitoxide

Byron commented:
You may also find this document about [diff-sliders](https://github.com/mhagger/diff-slider-tools/blob/master/README.md) interesting. It also details the tuning that Git has received to produce more 'likable' diffs, some of which `libgit2` seems to inherit. It seems like an art as well.

Right now, `imara-diff` has received no time in the direction of optimizing diffs for humans, and I think it would have to be a community effort if that should happen.

As this comes up all the time, I also feel that pressure is mounting, and more so the more people use `imara-diff` :).

blinxen commented:
I think that trying to improve `imara-diff` would be more interessting. I checked the `git` mailing list and found the initial patch that introduced what the [diff-sliders document](https://github.com/mhagger/diff-slider-tools/blob/master/README.md) proposes. I will check it out and try to implement something similiar for `imara-diff`. At the mean time we can keep this PR open since it does not have any urgency.

blinxen committed:
Run cargo fmt

blinxen committed:
Fix deprecation warnings after updating gix

blinxen committed:
Update expected results for git diffs

The diffing library that gitoxide uses does not have the same optimizations as libgit2 (yet)

blinxen committed:
Set author when setting up test git environment

blinxen committed:
Fix import order

blinxen committed:
Add changelog entry

blinxen commented:
Since I have not been able to work much on the `imara-diff` improvements lately, I updated this PR so that it at least passes all tests.

blinxen commented:
https://github.com/GitoxideLabs/gitoxide/pull/2011 just got merged and I tested the optimized diff locally and it works :D. I will updated this PR once a new version of `gix` is released. Then this PR will be open for review.

Byron commented:
I am sorry for my negligence and wishful thinking when merging the implementation despite it having only one test, but it does indeed not appear to work correctly.

Having put it into GitButler to get some real-world experience with it, here we have an example where both Git and GitButler (with the standard hunk visualization) agree:

<img width="1277" alt="expected" src="https://github.com/user-attachments/assets/92123f60-549f-4789-94d9-0f6bd4d25c09" />
<img width="1008" alt="expected-ui" src="https://github.com/user-attachments/assets/6d005698-aae6-4a35-8410-e7e727b9973c" />

However, when using the new hunk processing to get a Git-like diff, this happens:

<img width="1372" alt="wrong-with-git-diff-sink" src="https://github.com/user-attachments/assets/01ba1f1f-0914-48ff-861d-d0e9ee84c121" />

Of course it could be that the `ChangeKind` needs to be interpreted in some way, but even if so, I think it needs tests to make clear how this is working, the more the merrier.

For now I have removed the implementation from the codebase, but of course would be happy to see a follow-up PR that brings it back, along with tests, ideally even baseline tests where various expectations are produced by Git itself.

blinxen commented:
That is not so good :grimacing:

I took a quick peek and something with the scoring is weird with this diff. For some reason the implementation prefers moving up to line 132 over staying at line 141. I will look into it and also try to come up with a good test suite.

blinxen commented:
Already found the issue --> https://github.com/blinxen/gitoxide/commit/d7ab68e659df3cd99bc69fd89932575a28f4691c

>  Maybe one can bring in some tests from [the diff-slider repo](https://github.com/mhagger/diff-slider-tools/blob/master/README.md) to get more coverage.

That will be the next step

keith-hall commented:
Any new updates on this one? :)

blinxen commented:
>  What do you think? Too much work?

I don't see a benefit in this. I initially started this PR to reduce maintenance work on `bat` in Fedora. So if we are going to have the `gitoxide` feature then I would like to use it in Fedora. That however will make people create issues here that on Fedora they see a different diff output. Therefore I suggest to close this PR and maybe come back again later when the whole situation is better.

Byron commented:
This makes me sad, but I also have no solution for the diff-slider issue and the Git related modifications.
It's notable though that the version of this that `helix` is using was basically conjured up over night, with no tests that I know of.
And I don't think `gitoxide` will ever test that in production, so it's indeed better to hold back and see how the situation resolves itself.

Byron commented:
@blinxen Gitoxide now provides Git slider postprocessing for diffs, which seems to be ~95% accurate against baseline Git. You could probably experiment with it for a chance to bring this back.

Byron commented:
`gix` 0.82 comes with this slider post-processing. It might be worth giving this another try with particular attention to those slider-situations where post-processing should yield better outcomes.
